packages/synapse-react-client/src/components/QueryWrapper/QueryWrapper.tsx

```diff
@@ -47,56 +49,52 @@ export function QueryWrapper(props: QueryWrapperProps) {
     componentIndex,
     shouldDeepLink,
     onViewSharingSettingsClicked,
+    isRowSelectionVisible: isRowSelectionVisibleFromProps = false,
+    rowSelectionPrimaryKey: rowSelectionPrimaryKeyFromProps,
+    isInfinite = false,
   } = props
 
-  const [currentAsyncStatus, setCurrentAsyncStatus] = useState<
-    AsynchronousJobStatus<QueryBundleRequest, QueryResultBundle> | undefined
-  >(undefined)
+  // Must store requireConfirmationOnChange in state to break dependency cycle between useImmutableTableQuery and useTableRowsSelection
+  // TODO: consider combining these two hooks, or decoupling them in some other way
+  const [requireConfirmationOnChange, setRequireConfirmationOnChange] =
+    useState(false)
 
+  const immutableTableQueryResult = useImmutableTableQuery({
+    initQueryRequest,
+    shouldDeepLink,
+    componentIndex,
+    onQueryChange,
+    requireConfirmationOnChange,
+  })
   const {
     entityId,
     versionNumber,
     getInitQueryRequest,
     getLastQueryRequest,
     setQuery,
-    currentPage,
-    pageSize,
-    goToPage,
-    setPageSize,
     resetQuery,
     removeSelectedFacet,
     removeValueFromSelectedFacet,
     removeQueryFilter,
     removeValueFromQueryFilter,
-  } = useImmutableTableQuery({
-    initQueryRequest,
-    shouldDeepLink,
-    componentIndex,
-    onQueryChange,
-  })
+    onConfirmChange,
+    isConfirmingChange,
+    onCancelChange,
+  } = immutableTableQueryResult
 
   const lastQueryRequest = useMemo(() => {
     return getLastQueryRequest()
   }, [getLastQueryRequest])
 
-  const {
-    data: asyncJobStatus,
-    isLoading: queryIsLoading,
-    error,
-    isPreviousData: newQueryIsFetching,
-  } = useGetQueryResultBundleWithAsyncStatus(
-    lastQueryRequest,
-    {
-      // We use `keepPreviousData` because we don't want to clear out the current data when the query is modified via the UI
-      keepPreviousData: true,
-    },
-    setCurrentAsyncStatus,
-  )
-
-  const data = asyncJobStatus?.responseBody
+  const queryWrapperData = useQueryWrapperData(lastQueryRequest, isInfinite)
+  const { data, isLoadingNewBundle, error, currentAsyncStatus } =
+    queryWrapperData
```

Comment:
`useQueryWrapperData` abstracts whether we're fetching infinite or paginated data.